File: src/sc_neurocore/accel/julia/neurons/expif.jl
# SPDX-License-Identifier: AGPL-3.0-or-later
# Commercial license available
# © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
# © Code 2020–2026 Miroslav Šotek. All rights reserved.
# ORCID: 0009-0009-3560-0851
# Contact: www.anulum.li | protoscience@anulum.li
# SC-NeuroCore — Julia Fourcaud-Trocmé ExpIF recurrence

module ExpifAccel

export ExpIFNeuronState, reset!, simulate, simulate_complete, simulate_trace, step!

mutable struct ExpIFNeuronState
    v::Float64
    v_rest::Float64
    v_reset::Float64
    v_threshold::Float64
    v_rh::Float64
    delta_t::Float64
    tau::Float64
    dt::Float64
    refractory_period::Float64
    refractory_remaining::Float64
    source_profile::Bool
end

ExpIFNeuronState() = ExpIFNeuronState(
    -65.0,
    -65.0,
    -68.0,
    30.0,
    -59.9,
    3.48,
    10.0,
    0.02,
    0.0,
    0.0,
    false,
)

function _valid(state::ExpIFNeuronState, dt::Float64)
    all(
        isfinite,
        (
            state.v,
            state.v_rest,
            state.v_reset,
            state.v_threshold,
            state.v_rh,
            state.delta_t,
            state.tau,
            dt,
            state.refractory_period,
            state.refractory_remaining,
        ),
    ) && state.delta_t > 0.0 && state.tau > 0.0 && dt > 0.0 &&
        state.refractory_period >= 0.0 && state.refractory_remaining >= 0.0 &&
        state.refractory_remaining <= state.refractory_period &&
        state.v_threshold > state.v_rh && state.v < state.v_threshold &&
        state.v_rest < state.v_threshold && state.v_reset < state.v_threshold &&
        (
            !state.source_profile ||
            (
                state.v_rest == -65.0 &&
                state.v_reset == -68.0 &&
                state.v_threshold == -30.0 &&
                state.v_rh == -59.9 &&
                state.delta_t == 3.48 &&
                state.tau == 10.0 &&
                dt < 0.02 &&
                state.refractory_period == 1.7
            )
        )
end

function _rhs(state::ExpIFNeuronState, v::Float64, current::Float64)
    bounded_v = min(v, state.v_threshold)
    exp_term = state.delta_t * exp((bounded_v - state.v_rh) / state.delta_t)
    rhs = (-(bounded_v - state.v_rest) + exp_term + current) / state.tau
    isfinite(rhs) || throw(DomainError(rhs, "ExpIF derivative must remain finite"))
    rhs
end

function step!(state::ExpIFNeuronState, current::Float64 = 0.0; dt::Float64 = state.dt)
    _valid(state, dt) || throw(DomainError(state.v, "ExpIF state parameters are invalid"))
    isfinite(current) || throw(DomainError(current, "ExpIF input current must be finite"))

    if state.refractory_remaining > 0.0
        next_refractory = max(0.0, state.refractory_remaining - dt)
        state.dt = dt
        state.refractory_remaining = next_refractory
        state.v = state.v_reset
        return 0
    end

    k1 = _rhs(state, state.v, current)
    if state.source_profile
        predictor = state.v + dt * k1
        k2 = _rhs(state, predictor, current)
        next_v = state.v + 0.5 * dt * (k1 + k2)
    else
        k2 = _rhs(state, state.v + 0.5 * dt * k1, current)
        k3 = _rhs(state, state.v + 0.5 * dt * k2, current)
        k4 = _rhs(state, state.v + dt * k3, current)
        next_v = state.v + (dt / 6.0) * (k1 + 2.0 * k2 + 2.0 * k3 + k4)
    end
    isfinite(next_v) || throw(DomainError(next_v, "ExpIF update must remain finite"))

    state.dt = dt
    if next_v >= state.v_threshold
        state.v = state.v_reset
        state.refractory_remaining = state.refractory_period
        return 1
    end
    state.v = next_v
    0
end

function reset!(state::ExpIFNeuronState)
    state.v = state.v_rest
    state.refractory_remaining = 0.0
    nothing
end

function simulate_complete(
    v::Float64,
    v_rest::Float64,
    v_reset::Float64,
    v_threshold::Float64,
    v_rh::Float64,
    delta_t::Float64,
    tau::Float64,
    dt::Float64,
    refractory_period::Float64,
    refractory_remaining::Float64,
    source_profile::Bool,
    n_steps::Int,
    current::Float64,
)
    n_steps >= 0 || throw(ArgumentError("n_steps must be non-negative"))
    state = ExpIFNeuronState(
        v,
        v_rest,
        v_reset,
        v_threshold,
        v_rh,
        delta_t,
        tau,
        dt,
        refractory_period,
        refractory_remaining,
        source_profile,
    )
    _valid(state, dt) || throw(DomainError(v, "ExpIF state parameters are invalid"))
    isfinite(current) || throw(DomainError(current, "ExpIF input current must be finite"))
    voltage = Vector{Float64}(undef, n_steps)
    refractory = Vector{Float64}(undef, n_steps)
    events = Vector{UInt8}(undef, n_steps)
    spikes = 0
    for index in eachindex(voltage)
        event = step!(state, current)
        spikes += event
        voltage[index] = state.v
        refractory[index] = state.refractory_remaining
        events[index] = UInt8(event)
    end
    (
        voltage = voltage,
        refractory = refractory,
        events = events,
        spikes = spikes,
        vf = state.v,
        rf = state.refractory_remaining,
    )
end

function simulate_trace(args...)
    result = simulate_complete(args...)
    (
        trace = result.voltage,
        spikes = result.spikes,
        vf = result.vf,
        rf = result.rf,
    )
end

function simulate(n_steps::Int, current::Float64 = 0.0)
    state = ExpIFNeuronState()
    result = simulate_trace(
        state.v,
        state.v_rest,
        state.v_reset,
        state.v_threshold,
        state.v_rh,
        state.delta_t,
        state.tau,
        state.dt,
        state.refractory_period,
        state.refractory_remaining,
        state.source_profile,
        n_steps,
        current,
    )
    result.trace, result.spikes
end

end

File: src/sc_neurocore/accel/go/services/expif.go
// SPDX-License-Identifier: AGPL-3.0-or-later
// Commercial license available
// © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
// © Code 2020–2026 Miroslav Šotek. All rights reserved.
// ORCID: 0009-0009-3560-0851
// Contact: www.anulum.li | protoscience@anulum.li
// SC-NeuroCore — Go Fourcaud-Trocmé ExpIF recurrence

package services

import (
	"errors"
	"math"
)

var (
	ErrExpIFInvalidInput    = errors.New("expif input current must be finite")
	ErrExpIFInvalidState    = errors.New("expif state or parameters violate the runtime contract")
	ErrExpIFNonFiniteUpdate = errors.New("expif rk4 update must remain finite")
)

// ExpIFNeuronState holds the full deterministic and refractory state.
type ExpIFNeuronState struct {
	V                   float64
	VRest               float64
	VReset              float64
	VThreshold          float64
	VRh                 float64
	DeltaT              float64
	Tau                 float64
	Dt                  float64
	RefractoryPeriod    float64
	RefractoryRemaining float64
	SourceProfile       bool
}

// NewFourcaudTrocme2003ExpIF returns the fitted deterministic zero-noise profile.
func NewFourcaudTrocme2003ExpIF() *ExpIFNeuronState {
	state := NewExpIFNeuron()
	state.VThreshold = -30.0
	state.Dt = 0.01
	state.RefractoryPeriod = 1.7
	state.SourceProfile = true
	return state
}

// NewExpIFNeuron returns the historical SC RK4 compatibility defaults.
func NewExpIFNeuron() *ExpIFNeuronState {
	return &ExpIFNeuronState{
		V:                   -65.0,
		VRest:               -65.0,
		VReset:              -68.0,
		VThreshold:          30.0,
		VRh:                 -59.9,
		DeltaT:              3.48,
		Tau:                 10.0,
		Dt:                  0.02,
		RefractoryPeriod:    0.0,
		RefractoryRemaining: 0.0,
	}
}

// Valid reports whether every precondition of Step holds.
func (s *ExpIFNeuronState) Valid() bool {
	return finiteExpIF(
		s.V,
		s.VRest,
		s.VReset,
		s.VThreshold,
		s.VRh,
		s.DeltaT,
		s.Tau,
		s.Dt,
		s.RefractoryPeriod,
		s.RefractoryRemaining,
	) && s.DeltaT > 0.0 && s.Tau > 0.0 && s.Dt > 0.0 &&
		s.RefractoryPeriod >= 0.0 && s.RefractoryRemaining >= 0.0 &&
		s.RefractoryRemaining <= s.RefractoryPeriod && s.VThreshold > s.VRh &&
		s.V < s.VThreshold && s.VRest < s.VThreshold && s.VReset < s.VThreshold &&
		(!s.SourceProfile || (s.VRest == -65.0 && s.VReset == -68.0 &&
			s.VThreshold == -30.0 && s.VRh == -59.9 && s.DeltaT == 3.48 &&
			s.Tau == 10.0 && s.Dt < 0.02 && s.RefractoryPeriod == 1.7))
}

func finiteExpIF(values ...float64) bool {
	for _, value := range values {
		if math.IsNaN(value) || math.IsInf(value, 0) {
			return false
		}
	}
	return true
}

func (s *ExpIFNeuronState) rhs(v float64, current float64) float64 {
	boundedV := math.Min(v, s.VThreshold)
	arg := (boundedV - s.VRh) / s.DeltaT
	expTerm := s.DeltaT * math.Exp(arg)
	return (-(boundedV - s.VRest) + expTerm + current) / s.Tau
}

// Step advances one profile-selected Runge-Kutta update without partial mutation.
func (s *ExpIFNeuronState) Step(current float64) (int, error) {
	if !finiteExpIF(current) {
		return 0, ErrExpIFInvalidInput
	}
	if !s.Valid() {
		return 0, ErrExpIFInvalidState
	}

	if s.RefractoryRemaining > 0.0 {
		s.RefractoryRemaining = math.Max(0.0, s.RefractoryRemaining-s.Dt)
		s.V = s.VReset
		return 0, nil
	}

	k1 := s.rhs(s.V, current)
	var k2, k3, k4, nextV float64
	if s.SourceProfile {
		k2 = s.rhs(s.V+s.Dt*k1, current)
		nextV = s.V + 0.5*s.Dt*(k1+k2)
	} else {
		k2 = s.rhs(s.V+0.5*s.Dt*k1, current)
		k3 = s.rhs(s.V+0.5*s.Dt*k2, current)
		k4 = s.rhs(s.V+s.Dt*k3, current)
		nextV = s.V + (s.Dt/6.0)*(k1+2.0*k2+2.0*k3+k4)
	}
	if !finiteExpIF(k1, k2, k3, k4, nextV) {
		return 0, ErrExpIFNonFiniteUpdate
	}

	if nextV >= s.VThreshold {
		s.V = s.VReset
		s.RefractoryRemaining = s.RefractoryPeriod
		return 1, nil
	}
	s.V = nextV
	return 0, nil
}

// Reset restores resting voltage and clears the refractory hold.
func (s *ExpIFNeuronState) Reset() {
	s.V = s.VRest
	s.RefractoryRemaining = 0.0
}

// SimulateExpIFNeuron runs the default recurrence and returns a post-step trace.
func SimulateExpIFNeuron(nSteps int, current float64) ([]float64, int, error) {
	if nSteps < 0 {
		return nil, 0, ErrExpIFInvalidState
	}
	state := NewExpIFNeuron()
	trace := make([]float64, nSteps)
	spikes := 0
	for index := range trace {
		spike, err := state.Step(current)
		if err != nil {
			return nil, 0, err
		}
		spikes += spike
		trace[index] = state.V
	}
	return trace, spikes, nil
}

// SimulateExpIFComplete runs a candidate copy and returns aligned state/event rows.
// The supplied receiver is committed only after the complete batch succeeds.
func (s *ExpIFNeuronState) SimulateExpIFComplete(
	nSteps int,
	current float64,
) ([]float64, []float64, []uint8, error) {
	if nSteps < 0 || !finiteExpIF(current) || !s.Valid() {
		return nil, nil, nil, ErrExpIFInvalidState
	}
	candidate := *s
	voltage := make([]float64, nSteps)
	refractory := make([]float64, nSteps)
	events := make([]uint8, nSteps)
	for index := 0; index < nSteps; index++ {
		event, err := candidate.Step(current)
		if err != nil {
			return nil, nil, nil, err
		}
		voltage[index] = candidate.V
		refractory[index] = candidate.RefractoryRemaining
		events[index] = uint8(event)
	}
	*s = candidate
	return voltage, refractory, events, nil
}
